util/Pending.h

```diff
@@ -38,62 +38,84 @@ namespace Pending {
 
             if (status == std::future_status::deferred) {
                 ErrorLogger() << "Pending parse is unable to handle deferred future.";
                 throw "deferred future not handled";
             }
-
+            DebugLogger() << "WaitForPendingUnlocked another wait_for round";
         } while (status != std::future_status::ready);
 
         try {
             // multiple threads might be waiting but not care about the results
             if (do_not_care_about_result) {
-                if (pending && pending->pending->valid()) {
-                    DebugLogger() << "Dont care for result of parsing \"" << pending->filename << "\". Have to get() once to release shared state in pending future.";
-                    pending->pending->get(); // needs to be called once to release state
+                if (pending.pending->valid()) {
+                    DebugLogger() << "Dont care for result of parsing \"" << pending.filename << "\". Have to get() once to release shared state in pending future.";
+                    pending.pending->get(); // needs to be called once to release state
                 }
-                DebugLogger() << "Dont care for result of parsing \"" << pending->filename << "\". Was already released.";
-                pending = boost::none;
+                DebugLogger() << "Dont care for result of parsing \"" << pending.filename << "\". Was already released.";
                 return boost::none;
             }
-            DebugLogger() << "Retrieve result of parsing \"" << pending->filename << "\".";
-            auto x = std::move(pending->pending->get());
-            pending = boost::none;
+            DebugLogger() << "Retrieve result of parsing \"" << pending.filename << "\".";
+            auto x = std::move(pending.pending->get());
+            DebugLogger() << "Retrieved result of parsing \"" << pending.filename << "\".";
             return std::move(x);
         } catch (const std::exception& e) {
-            ErrorLogger() << "Parsing of \"" << pending->filename << "\" failed with error: " << e.what();
-            pending = boost::none;
+            ErrorLogger() << "Parsing of \"" << pending.filename << "\" failed with error: " << e.what();
         }
 
         return boost::none;
     }
+    /** Wait for the \p pending parse to complete.  Set pending to boost::none
+        and return the parsed T. Destroys the shared state in the wrapped std::future.
+        Return boost::none on errors.*/
+    template <typename T>
+    boost::optional<T> WaitForPending(boost::optional<Pending<T>>& pending, bool do_not_care_about_result = false) {
+        if (!pending)
+            return boost::none;
+        std::lock_guard<std::mutex> lock(pending->m_mutex);
+        if (!pending || !(pending->pending)) {
+            return boost::none;
+        }
+        if (auto tt = WaitForPendingUnlocked(std::move(*pending), do_not_care_about_result)) {
+            pending = boost::none; // XXX
```

Comment:
`XXX`?

Reply:
Setting pending to boost::none probably could happen in WaitForPendingUnlocked, instead of in the wrapping WaitForPending and SwapPending. Beforehand it was in WaitForPending and SwapPending was wrapping it. I guess it is mostly matter of style where to do that, what would you prefer?